# Answering Business Questions using SQL

## Introduction

In this guided project, we'll use the Chinook database that we've used in earlier missions of this course in Dataquest. The Chinook database is provided as a SQLite database file called __chinook.db__. A copy of the database schema is below:

![Chinook database schema](https://s3.amazonaws.com/dq-content/191/chinook-schema.svg)

Let's start by loading the database and the code tu run SQL in this environment, and then reviewing the tables.

In [1]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db

'Connected: None@chinook.db'

In [2]:
%%sql
SELECT
    name,
    type
FROM sqlite_master
WHERE type IN ("table","view");

Done.


[('album', 'table'),
 ('artist', 'table'),
 ('customer', 'table'),
 ('employee', 'table'),
 ('genre', 'table'),
 ('invoice', 'table'),
 ('invoice_line', 'table'),
 ('media_type', 'table'),
 ('playlist', 'table'),
 ('playlist_track', 'table'),
 ('track', 'table')]

In [3]:
%%sql
SELECT *
FROM album
LIMIT 10;

Done.


[(1, 'For Those About To Rock We Salute You', 1),
 (2, 'Balls to the Wall', 2),
 (3, 'Restless and Wild', 2),
 (4, 'Let There Be Rock', 1),
 (5, 'Big Ones', 3),
 (6, 'Jagged Little Pill', 4),
 (7, 'Facelift', 5),
 (8, 'Warner 25 Anos', 6),
 (9, 'Plays Metallica By Four Cellos', 7),
 (10, 'Audioslave', 8)]

In [4]:
%%sql
SELECT *
FROM customer
LIMIT 10;

Done.


[(1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3),
 (2, 'Leonie', 'Köhler', None, 'Theodor-Heuss-Straße 34', 'Stuttgart', None, 'Germany', '70174', '+49 0711 2842222', None, 'leonekohler@surfeu.de', 5),
 (3, 'François', 'Tremblay', None, '1498 rue Bélanger', 'Montréal', 'QC', 'Canada', 'H2G 1A7', '+1 (514) 721-4711', None, 'ftremblay@gmail.com', 3),
 (4, 'Bjørn', 'Hansen', None, 'Ullevålsveien 14', 'Oslo', None, 'Norway', '0171', '+47 22 44 22 22', None, 'bjorn.hansen@yahoo.no', 4),
 (5, 'František', 'Wichterlová', 'JetBrains s.r.o.', 'Klanova 9/506', 'Prague', None, 'Czech Republic', '14700', '+420 2 4172 5555', '+420 2 4172 5555', 'frantisekw@jetbrains.com', 4),
 (6, 'Helena', 'Holý', None, 'Rilská 3174/6', 'Prague', None, 'Czech Republic', '14300', '+420 2 4177 0449', None, 'hholy@gmail.com', 5),
 (7, 'Astrid', 'Gruber', None, 'Rotenturmstraße 4, 1010 Innere Stadt', 'Vienne', None, 'Austria', '1010', '+43 01 5134505', None, 'astrid.gruber@apple.at', 5),
 (8, 'Daan', 'Peeters', None, 'Grétrystraat 63', 'Brussels', None, 'Belgium', '1000', '+32 02 219 03 03', None, 'daan_peeters@apple.be', 4),
 (9, 'Kara', 'Nielsen', None, 'Sønder Boulevard 51', 'Copenhagen', None, 'Denmark', '1720', '+453 3331 9991', None, 'kara.nielsen@jubii.dk', 4),
 (10, 'Eduardo', 'Martins', 'Woodstock Discos', 'Rua Dr. Falcão Filho, 155', 'São Paulo', 'SP', 'Brazil', '01007-010', '+55 (11) 3033-5446', '+55 (11) 3033-4564', 'eduardo@woodstock.com.br', 4)]

In [5]:
%%sql
SELECT *
FROM invoice
LIMIT 10;

Done.


[(1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84),
 (2, 30, '2017-01-03 00:00:00', '230 Elgin Street', 'Ottawa', 'ON', 'Canada', 'K2P 1L7', 9.9),
 (3, 40, '2017-01-05 00:00:00', '8, Rue Hanovre', 'Paris', 'None', 'France', '75002', 1.98),
 (4, 18, '2017-01-06 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 7.92),
 (5, 27, '2017-01-07 00:00:00', '1033 N Park Ave', 'Tucson', 'AZ', 'USA', '85719', 16.83),
 (6, 31, '2017-01-10 00:00:00', '194A Chain Lake Drive', 'Halifax', 'NS', 'Canada', 'B3S 1C5', 1.98),
 (7, 49, '2017-01-12 00:00:00', 'Ordynacka 10', 'Warsaw', 'None', 'Poland', '00-358', 10.89),
 (8, 59, '2017-01-13 00:00:00', '3,Raj Bhavan Road', 'Bangalore', 'None', 'India', '560001', 9.9),
 (9, 18, '2017-01-18 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 8.91),
 (10, 31, '2017-01-18 00:00:00', '194A Chain Lake Drive', 'Halifax', 'NS', 'Canada', 'B3S 1C5', 1.98)]

## Top Genres
Suppose that the Chinook record store has just signed a deal with a new record label, and we've been tasked with selecting the first three albums that will be added to the store, from a list of four. All four albums are by artists that don't have any tracks in the store right now - we have the artist names, and the genre of music they produce:

|      Artist name     |  Genre  |
|:--------------------:|:-------:|
|         Regal        | Hip-Hop |
|       Red Tone       |   Punk  |
| Meteor and the Girls |   Pop   |
|    Slim Jim Bites    |  Blues  |


The record label specializes in artists from the USA, and they have given Chinook some money to advertise the new albums in the USA, so we're interested in finding out which genres sell the best in the USA.

In [6]:
%%sql

WITH "top_tracks_usa" AS (SELECT
                        il.track_id track_id,
                        SUM(il.quantity) total
                    FROM invoice i
                    INNER JOIN customer c
                        ON c.customer_id = i.customer_id
                    INNER JOIN invoice_line il
                        ON i.invoice_id = il.invoice_id
                    WHERE c.country = 'USA'
                    GROUP BY il.track_id
                    ORDER BY total DESC),
"top_genres_usa" AS (SELECT
                            SUM(ttu.total) total_genre,
                            g.name genre_name
                        FROM top_tracks_usa ttu
                        INNER JOIN track t
                            ON t.track_id = ttu.track_id
                        INNER JOIN genre g
                            ON g.genre_id = t.genre_id
                        GROUP BY genre_name
                        ORDER BY total_genre DESC)
SELECT tgu.genre_name genre,
        tgu.total_genre tracks_sold,
        ROUND((CAST(tgu.total_genre AS Float)
               / (SELECT SUM(total_genre)
                  FROM top_genres_usa))*100,2) percentage_sold
FROM top_genres_usa tgu
GROUP BY genre
ORDER BY tracks_sold DESC;

Done.


[('Rock', 561, 53.38),
 ('Alternative & Punk', 130, 12.37),
 ('Metal', 124, 11.8),
 ('R&B/Soul', 53, 5.04),
 ('Blues', 36, 3.43),
 ('Alternative', 35, 3.33),
 ('Latin', 22, 2.09),
 ('Pop', 22, 2.09),
 ('Hip Hop/Rap', 20, 1.9),
 ('Jazz', 14, 1.33),
 ('Easy Listening', 13, 1.24),
 ('Reggae', 6, 0.57),
 ('Electronica/Dance', 5, 0.48),
 ('Classical', 4, 0.38),
 ('Heavy Metal', 3, 0.29),
 ('Soundtrack', 2, 0.19),
 ('TV Shows', 1, 0.1)]

We can see above the top genres in Chinook store. More than half of the top tracks are in Rock genre, followed up by Alternative & Punk and Metal (both lightly above 10 %) and then  R&B/Soul. Other genres are below 5 % of the total tracks.
Reviewing the available new artists, _Red Tone_(Punk) and _Slim Jim Bites_(Blues) could be the better choice, while _Regal_ and _Meteor and the Girls_ are in not-so-popular genres (Pop and Hip-Hop).

## Analyzing Employee Sales Performance
Each customer for the Chinook store gets assigned to a sales support agent within the company when they first make a purchase. We can analyze the purchases of customers belonging to each employee to see if any sales support agent is performing either better or worse than the others.

In [7]:
%%sql

SELECT
    e.first_name || " " || e.last_name employee_name,
    e.title title,
    e.hire_date hire_date,
    ROUND(SUM(i.total),2) total_purchased,
    e2.first_name || " " || e2.last_name supervisor_name,
    e2.title supervisor_title
FROM customer c
INNER JOIN employee e
    ON e.employee_id = c.support_rep_id
INNER JOIN invoice i
    ON i.customer_id = c.customer_id
INNER JOIN employee e2
    ON e.reports_to = e2.employee_id
GROUP BY support_rep_id
ORDER BY total_purchased DESC;

Done.


[('Jane Peacock', 'Sales Support Agent', '2017-04-01 00:00:00', 1731.51, 'Nancy Edwards', 'Sales Manager'),
 ('Margaret Park', 'Sales Support Agent', '2017-05-03 00:00:00', 1584.0, 'Nancy Edwards', 'Sales Manager'),
 ('Steve Johnson', 'Sales Support Agent', '2017-10-17 00:00:00', 1393.92, 'Nancy Edwards', 'Sales Manager')]

We can see from the table above that Jane Peacock is the top seller, and also the oldest sales agent. We can see a difference of about 10 % between Jane and Margaret, the second seller, wich is 13 % above Steve Johnson, the newest employee in this area. There is 1 month difference between Jane and Margaret's hire date, and 6 months between Jane and Steve. This migth indicate, that Steve is the lowest seller, but also the newest, so his growing rythm might be outstanding.

## Analyzing sales by country

We can analyse the sales data for customers from each different country by taking the country value from the __customers__ table, and ignore the country from the billing address in the __invoice__ table.

We'll try to find:
- total number of customers
- total value of sales
- average value of sales per customer
- average order value

Because there are a number of countries with only one customer, those will be classified  as "Other".

In [8]:
%%sql

WITH countries AS
    (SELECT
           CASE WHEN (SELECT count(*)
                     FROM customer
                     WHERE country = c.country) = 1
                THEN "Other"
               ELSE c.country
           END AS country,
           c.customer_id,
           il.*
    FROM invoice_line il
    INNER JOIN invoice i
       ON i.invoice_id = il.invoice_id
    INNER JOIN customer c
       ON c.customer_id = i.customer_id)
    
SELECT
    country,
    total_customers,
    total_sales,
    avg_per_customer,
    order_average
FROM (SELECT
        country,
        COUNT(DISTINCT customer_id) total_customers,
        ROUND(SUM(unit_price),2) total_sales,
        ROUND(SUM(unit_price) / count(distinct customer_id),2) avg_per_customer,
        ROUND(SUM(unit_price) / count(distinct invoice_id),2) order_average,
        CASE
            WHEN country = "Other" THEN 1
            ELSE 0
        END AS sort
    FROM countries
    GROUP BY country
    ORDER BY sort ASC, total_sales DESC);

Done.


[('USA', 13, 1040.49, 80.04, 7.94),
 ('Canada', 8, 535.59, 66.95, 7.05),
 ('Brazil', 5, 427.68, 85.54, 7.01),
 ('France', 5, 389.07, 77.81, 7.78),
 ('Germany', 4, 334.62, 83.66, 8.16),
 ('Czech Republic', 2, 273.24, 136.62, 9.11),
 ('United Kingdom', 3, 245.52, 81.84, 8.77),
 ('Portugal', 2, 185.13, 92.57, 6.38),
 ('India', 2, 183.15, 91.58, 8.72),
 ('Other', 15, 1094.94, 73.0, 7.45)]

## Albums or individual tracks?

The Chinook store is setup in a way that allows customer to make purchases in one of the two ways:

- purchase a whole album
- purchase a collection of one or more individual tracks.

The store does not let customers purchase a whole album, and then add individual tracks to that same purchase (unless they do that by choosing each track manually). When customers purchase albums they are charged the same price as if they had purchased each of those tracks separately.

Management are currently considering changing their purchasing strategy to save money. The strategy they are considering is to purchase only the most popular tracks from each album from record companies, instead of purchasing every track from an album.

We need to find out what percentage of purchases are individual tracks vs whole albums, so that management can use this data to understand the effect this decision might have on overall revenue. It is important to notice that we will drop albums with 1 or 2 tracks only, because they can be interpreted as coincidences in individual purchases.

In [9]:
%%sql

WITH invoice_first_track AS (
                        SELECT
                             il.invoice_id invoice_id,
                             MIN(il.track_id) first_track_id
                         FROM invoice_line il
                         GROUP BY 1
                         )

SELECT
    purchase,
    COUNT(invoice_id) total_invoices,
    ROUND(CAST(count(invoice_id) AS FLOAT) / 
            (SELECT COUNT(*) FROM invoice),4)*100 percent
FROM (SELECT
        ifs.*,
        CASE
            WHEN (SELECT t.track_id
                  FROM track t
                  WHERE t.album_id = (SELECT t2.album_id
                                      FROM track t2
                                      WHERE t2.track_id = ifs.first_track_id) 
      EXCEPT 

      SELECT il2.track_id FROM invoice_line il2
      WHERE il2.invoice_id = ifs.invoice_id) IS NULL
             AND (SELECT il2.track_id
                  FROM invoice_line il2
                  WHERE il2.invoice_id = ifs.invoice_id

                  EXCEPT 

                  SELECT t.track_id
                  FROM track t
                  WHERE t.album_id = (SELECT t2.album_id
                                      FROM track t2
                                      WHERE t2.track_id = ifs.first_track_id)) IS NULL THEN "album"
             ELSE "tracks"
         END AS "purchase"
                 FROM invoice_first_track ifs)
GROUP BY purchase;

Done.


[('album', 114, 18.57), ('tracks', 500, 81.43)]

We can see that the most popular buying method is by tracks, wito more than 80% of purchases. Only about 20% of total purchases are for complete albums, so this migth result in 18.57% lost in revenue.